homeassistant/components/nobo_hub/climate.py

```diff
@@ -0,0 +1,326 @@
+    _attr_temperature_unit = TEMP_CELSIUS
+
+    def __init__(
+        self, zone_id, hub: nobo, command_off_id, command_on_id, override_type
+    ):
+        """Initialize the climate device."""
+        self._id = zone_id
+        self._nobo = hub
+        self._attr_unique_id = hub.hub_serial + ":" + zone_id
+        self._attr_name = hub.zones[self._id][ATTR_NAME]
+        self._attr_hvac_mode = HVAC_MODE_AUTO
+        self._command_off_id = command_off_id
+        self._command_on_id = command_on_id
+        self._override_type = override_type
+        self._can_turn_off = (
+            self._command_on_id is not None and self._command_off_id is not None
+        )
+
+        # Register for callbacks before initial update to avoid race condition.
+        hub.register_callback(self._after_update)
+        self.update()
+
+    @property
+    def hvac_modes(self):
+        """Return the list of available operation modes."""
+        # Only enable off-command if on- and off-command exists for this zone:
+        if self._can_turn_off:
+            return HVAC_MODES
+        return HVAC_MODES_WITHOUT_OFF
+
+    async def async_set_hvac_mode(self, hvac_mode):
+        """Set HVAC mode to comfort(HEAT) or back to normal(AUTO)."""
+        if hvac_mode == HVAC_MODE_AUTO:
+            await self.async_set_preset_mode(PRESET_NONE)
+            self._attr_hvac_mode = hvac_mode
+        elif hvac_mode == HVAC_MODE_HEAT:
+            await self.async_set_preset_mode(PRESET_COMFORT)
+            self._attr_hvac_mode = hvac_mode
+
+        if self._can_turn_off:
+            if hvac_mode == HVAC_MODE_OFF:
+                await self.async_set_preset_mode(PRESET_NONE)
+                self._attr_hvac_mode = hvac_mode
+                await self._nobo.async_update_zone(
+                    self._id, week_profile_id=self._command_off_id
+                )  # Change week profile to OFF
+                _LOGGER.debug(
+                    "Turned off heater %s '%s' by switching to week profile %s",
+                    self._id,
+                    self._attr_name,
+                    self._command_off_id,
+                )
+            else:
+                await self._nobo.async_update_zone(
+                    self._id, week_profile_id=self._command_on_id
+                )  # Change week profile to normal for this zone
+                _LOGGER.debug(
+                    "Turned on heater %s '%s' by switching to week profile %s",
+                    self._id,
+                    self._attr_name,
+                    self._command_on_id,
+                )
+            # When switching between AUTO and OFF an immediate update does not work (the Nobø API seems to answer with old values), but it works if we add a short delay:
+            await asyncio.sleep(0.5)
+        elif hvac_mode == HVAC_MODE_OFF:
+            _LOGGER.warning(
+                "User tried to turn off zone %s '%s', but this is not configured so this should be impossible",
```

Comment:
The logic for this could get simpler if at the top of the function there is something like `if hvac_mode not in self.hvac_modes` and it returns and logs a warning. It generalized a bit to also protect against other types of "impossible" cases.

Reply:
You may consider setting `_attr_hvac_modes` in the constructor (just append HVAC_MODE_OFF to the list when needed), and remove `_can_turn_off` and the separate _WITHOUT_OFF list.

Reply:
Thanks! Good idea - I'll do this.

Reply:
Fixed.